torchmultimodal/models/mdetr/model.py

```diff
@@ -257,3 +260,236 @@ def mdetr_resnet101(
+    answer_attr_head = nn.Linear(hidden_dim, 403)
+    answer_rel_head = nn.Linear(hidden_dim, 1594)
+    answer_global_head = nn.Linear(hidden_dim, 111)
+    answer_cat_head = nn.Linear(hidden_dim, 678)
+    heads = nn.ModuleDict(
+        {
+            "answer_type": answer_type_head,
+            "answer_obj": answer_obj_head,
+            "answer_rel": answer_rel_head,
+            "answer_attr": answer_attr_head,
+            "answer_cat": answer_cat_head,
+            "answer_global": answer_global_head,
+        }
+    )
+    return MultiHead(heads)
+
+
+class ContrastiveEmbeddingsOutput(NamedTuple):
+    query_embeddings: Tensor
+    token_embeddings: Tensor
+
+
+class MDETRVQAOutput(NamedTuple):
+    model_output: MDETRModelOutput
+    vqa_preds: Dict[str, Tensor]
+    contrastive_embeddings: ContrastiveEmbeddingsOutput
+
+
+class MDETRForVQA(nn.Module):
+    def __init__(
+        self,
+        model: MDETR,
+        vqa_heads: MultiHead,
+        contrastive_alignment_image_projection: nn.Module,
+        contrastive_alignment_text_projection: nn.Module,
+    ):
+        super().__init__()
+        self.model = model
+        self.vqa_heads = vqa_heads
+        if self.model.extra_query_embeddings is None:
+            raise ValueError("MDETRForVQA requires extra query embeddings ")
+        if self.model.extra_query_embeddings.num_embeddings != len(
+            self.vqa_heads.heads.keys()
+        ):
+            raise ValueError("Number of heads must match number of QA embeddings")
+
+        self.contrastive_alignment_image_projection = (
+            contrastive_alignment_image_projection
+        )
+        self.contrastive_alignment_text_projection = (
+            contrastive_alignment_text_projection
+        )
+
+    def forward(
+        self,
+        images: List[Tensor],
+        text: List[Tensor],
+    ) -> MDETRVQAOutput:
+        # Calculate MDETR model outputs
+        model_output = self.model(images, text)
+
+        final_hidden_state = model_output.transformer_output.decoder_hidden_states[-1]
+
+        # Perform projections for contrastive alignment loss.
+        contrastive_query_embeddings = F.normalize(
+            self.contrastive_alignment_image_projection(final_hidden_state),
+            p=2,
+            dim=-1,
+        )
+        contrastive_token_embeddings = F.normalize(
+            self.contrastive_alignment_text_projection(
+                model_output.transformer_output.text_memory
+            ).transpose(0, 1),
+            p=2,
+            dim=-1,
+        )
+        contrastive_outputs = ContrastiveEmbeddingsOutput(
+            contrastive_query_embeddings, contrastive_token_embeddings
+        )
+
+        # Apply VQA heads to get answer predictions
+        answer_preds = self.vqa_heads(model_output.extra_embeddings.transpose(0, 1))
+
+        return MDETRVQAOutput(model_output, answer_preds, contrastive_outputs)
+
+
+def mdetr_for_vqa(
```

Comment:
Consider adding an option to load from checkpoint to the builder (e.g., omnivore.py)?
From a user's perspective, I'd probably prefer a clean API such as `build_model(pretrained=True)`.

RE: this block in `eval.py`.
https://github.com/facebookresearch/multimodal/blob/9b3d1a17de0d9b20f9e83db423b24e26a4363e57/examples/mdetr/vqa_eval.py#L156-L163

Reply:
Yeah this is still a to-do, I am planning to address separately if that's good with you

Reply:
@langong347 Actually after thinking about this a bit more, I am not sure if there will be a clean approach to the checkpoint loading API for MDETR. There are a couple reasons for this. 

1) The full MDETR model checkpoint contains both model weights and EMA model weights. We could modify the builder to return both model and EMA model, but I don't think this is the right approach. (For instance for inference we only use the EMA model, so we would not be able to use it then.) So otherwise we will have to pass a key to e.g. the `load_module_from_url` method. (I think `PretrainedMixin` already supports this, but iirc we are thinking of deprecating it?) 

2) MDETR contains a handful of configurable components, meaning that there is not really a canonical checkpoint for a given task. For instance, if we are loading a checkpoint to do VQA eval we expect VQA heads in the state dict, but if we are loading a checkpoint to do VQA fine-tuning we should load the pretrained checkpoint directly, constructing the VQA heads ourselves. There is also the question of contrastive projections, which can be used or omitted depending on whether we are performing training or not. Of course we can set `strict=False` to allow for each of these cases, but then there is a higher likelihood that a user playing around with various configurations would break something and not realize it. 

Hopefully this makes sense, please let me know if any part of it is unclear.

Reply:
>MDETR contains a handful of configurable components, meaning that there is not really a canonical checkpoint for a given task. 

It sounds like your current checkpoint files are the superset of what you want for different purposes? 
Technically, is it possible to checkpoint different parts of the model separately?

Reply:
yeah i think we need to add the official ckpt loading option here (which u can hack around as we did for omnivore and flava). adding the key option to the function is reasonable too

Reply:
Regarding checkpointing different parts of the model separately, it's technically possible. But I'm not sure it will be intuitive for the end user. Let me give a bit of background first.

The existing model checkpoints look like `{"model": model_weights, "model_ema": model_ema_weights}`, where both model_weights and model_ema_weights have the same set of keys. Aside from the core MDETR model weights, these include
- VQA head weights for the fine-tuned VQA checkpoint, and
- Contrastive projections for the pretrained checkpoint.

(Note that `MDETRForVQA` can be used for either evaluation or fine-tuning, which means that our `MDETRForVQAClass` may load either one of these checkpoints.)

So how should we split the checkpoint? I think there are two ways to do this.

1) Create a separate checkpoint file for each of `model` and `model_ema`. 
2) Create a separate checkpoint file for each of e.g. `core_model_weights`, `contrastive_projections`, and `vqa_heads`, then mix and match as needed.

I don't like (1) because it can be solved already by adding the key option to `load_module_from_url`. Then there is no extra confusion to the user and we still have all the weights in one place.

For (2), I think it is very counterintuitive. It is not really a standard kind of encoder/head split, and as a user I would be very confused about why a couple projections necessitate their own separate checkpoint. There is also the question of how this would mix with (1) (since each of these is nested under the model/model_ema keys).

Assuming that we agree we don't want to split the checkpoint (maybe a big assumption 🙂), the remaining question is how to handle the case where the weights don't line up perfectly. This can happen if e.g. the user loads `MDETRForVQA` with pretrained weights to do fine-tuning on the GQA dataset (since the model class will have VQA heads, which won't be present in the checkpoint). In this case, I think we can just load with `strict=False`. We have control over the checkpoints and builder functions so it is unlikely that users will be able modify things in such a way that `strict=False` will mask some other checkpoint loading error. I think [FLAVA](https://github.com/facebookresearch/multimodal/blob/679f3596e4c44b483c68d4023b24e3c7f77292b3/torchmultimodal/models/flava/flava_model.py#L565) is also doing this for a similar case. This way we don't have to add a bunch of slightly different checkpoints or a bunch of branching logic to the model classes or builders.

Let me know if this makes sense and if you have any concerns with the proposed approach @langong347 @ankitade.

Reply:
Not breaking the checkpoint file into core_model, projection and vqa_heads to avoid confusing the user. This makes sense to me since you have more pieces than just `encoder` and `head`.

>I don't like (1) because it can be solved already by adding the key option to load_module_from_url. Then there is no extra confusion to the user and we still have all the weights in one place.

Do you mean `mdtr_model.ckpt` and `mdtr_model_ema.ckpt` are confusing? That might be a personal choice I think. If we keep them in one file, you'll need an additional option in `load_module_from_url` to branch out? This could be too specific.

Reply:
Sorry for my delayed response as well 🙂 . Yeah I agree that there is some aspect of personal preference here. I think the branching should not be too difficult, it would be the same as what we already had in [PretrainedMixin](https://github.com/facebookresearch/multimodal/blob/8e99bf08a545acd99498bb69e768510db7069077/torchmultimodal/utils/common.py#L149). 

I also think the answer here depends on how the potentially separated weights would be used. For instance, in Omnivore the weights for encoder and heads are separated, because the user may want to plug in each one of these separately with other components, and so modularity becomes important. But for EMA it is not about plugging in as much as it is about train vs eval (in train we will want both model and EMA weights, in eval we only want EMA weights). So imo the value of keeping separate weights for model and EMA is not very high and it is better to just filter the state dict if calling from an eval loop.